### PR TITLE
Issue#250

### PR DESCRIPTION
**Pull request by charlie0605**
HTTP ERROR 422 is fixed, Rspec test is written and all pass

duplicate with space:
image
image

adding tag, no space but with dash
image
image

empty name:
image
image

### CODE REVIEW

File: spec/controllers/tags_controller_spec.rb [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+          @controller = BackdropsController.new
+
+          post :create, :params => { :backdrop => { :name => "bdTest", :tag => {name: "test space"} } }
+
+          expect{
+            post :create, :params => { :backdrop => { :name => "bdTest", :tag => {name: "test space"} } }
+          }.to change{Tag.count}.by(0)
+        end
+
+        it "backdrop should only have one tag" do
```
Comment (thongteav) [inline spec/controllers/tags_controller_spec.rb]: This should be sound

File: spec/controllers/tags_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+          @controller = BackdropsController.new
+
+          post :create, :params => { :backdrop => { :name => "bdTest", :tag => {name: "test"} } }
+
+          expect{
+            post :create, :params => { :backdrop => { :name => "bdTest", :tag => {name: "test"} } }
+          }.to change{Tag.count}.by(0)
+        end
+
+        it "backdrop should only have one tag" do
```
Comment (thongteav) [inline spec/controllers/tags_controller_spec.rb]: this should be sound too

File: app/controllers/tags_controller.rb [hunk 3]
```diff
@@ -71,6 +81,10 @@ def tag_params
       params.require(:tag).permit([:name])
     end
 
+    def format_params
+      params[:tag][:name].strip.downcase.gsub(/[^a-z0-9\s\-]/, '').gsub(/\s/, '-')
+    end
+
```
Comment (rohrlach) [inline app/controllers/tags_controller.rb]: This already exists as a pre-validation hook on the ```Tag``` model.  

We don't want code repeated anywhere.

Perhaps (as suggested) you should do this instead:
```
@tag = Tag.new(tag_params)
@tag.validate
```

Simply calling ```.validate``` will actually amend the name as needed (If you take a look at the code)

Comment (rohrlach) [reply app/controllers/tags_controller.rb]: Actually... Come to think of it, the cleaner way to do this is to change ```Tag``` such that the name standardization is ```after_initialize``` (Which is what I think I MEANT to do initially)

Comment (charlie0605) [reply app/controllers/tags_controller.rb]: Tag.new(tag_params) standardizes the name already doesn it? I have removed the extra function and used the @tag.name for the param when finding the tag name, is that a better way of doing it?

Comment (rohrlach) [reply app/controllers/tags_controller.rb]: Not currently, no.  line 2 of ```Tag``` is:
```
before_validation :standardize_name
```

Consider the following:
```
def create
    @tag = Tag.find_by_name(params[:tag][:name])
    @tag = Tag.new(tag_params) unless @tag.present?    # after_initialization is here
    if @tag.new_record?
      if @tag.save    # before_validation is here
...
```

If we were to change ```Tag``` such that line 2 reads: ```after_initialization :standardize_name``` we could then do the following:
```
def create
    @tag = Tag.new(tag_params) unless Tag.find_by_name(Tag.new(tag_params)).exists?
...
```

And all other code can remain the same.

**Comment by charlie0605** [comment]
@rohrlach Tag.find_by_name(Tag.new(tag_params)).exists? doesnt have the exists method, I have changed the code to Tag.exists?(name: @tag.name)
Is that fine?

File: app/controllers/tags_controller.rb [hunk 4]
```diff
@@ -16,27 +16,39 @@ def new
...
   def create
-    @tag = Tag.find_by_name(params[:tag][:name])
-    @tag = Tag.new(tag_params) unless @tag.present?
-    if @tag.new_record?
-      if @tag.save
-        flash[:success] = 'Tag created.'
-      else
-        flash.now[:danger] = 'Failed to create tag'
-        render :new
+    unless params[:tag][:name].blank?
```
Comment (rohrlach) [inline app/controllers/tags_controller.rb]: Why are you checking this here?  the model already checks this during validation

File: app/controllers/tags_controller.rb [hunk 5]
```diff
@@ -16,27 +16,39 @@ def new
...
+          end
         end
+        redirect_to @perspective
+      else
+        redirect_to tags_path
       end
-      redirect_to @perspective
     else
-      redirect_to @tag
+      flash[:danger] = "Name cannot be blank!"
```
Comment (rohrlach) [inline app/controllers/tags_controller.rb]: Again, the model takes care of this for you - You shouldn't do it in the controller.

Comment (charlie0605) [reply app/controllers/tags_controller.rb]: I tried not having it in there but it doesn't flash any error, but it gives me 500 Internal Server Error

Comment (rohrlach) [reply app/controllers/tags_controller.rb]: We'll discuss this evening at the meeting

**Comment by rohrlach** [comment]
Moving this to another branch - @charlie0605 I'll work with you on this.

**Pull request closed without merging**